File: WC3/C++_homework/Basic/Complex/complex.cpp
#include <iostream>
#include "Complex.h"

Complex::Complex(double real, double imaginary) {
    this->real = real;
    this->imaginary = imaginary;
}
Complex::~Complex() {}

double Complex::getReal() const {
    return this->real;
}
double Complex::getImaginary() const {
    return this->imaginary;
}

bool Complex::operator==(const Complex& other) const {
    return this->real == other.real && this->imaginary == other.imaginary;
}
bool Complex::operator!=(const Complex& other) const {
    return !((*this) == other);
}
void Complex::operator+=(const Complex& other) {
    this->real += other.real;
    this->imaginary += other.imaginary;
}
void Complex::operator-=(const Complex& other) {
    this->real -= other.real;
    this->imaginary -= other.imaginary;
}
Complex Complex::operator+(const Complex& other) const {
    return Complex(this->real+other.real, this->imaginary+other.imaginary);
}
Complex Complex::operator-(const Complex& other) const {
    return Complex(this->real-other.real, this->imaginary-other.imaginary);
}
Complex Complex::operator*(const Complex& other) const {
    return Complex(this->real * other.real - this->imaginary * other.imaginary, this->imaginary * other.real + this->real * other.imaginary);
}
std::ostream& operator<<(std::ostream& out, const Complex& complex) {
    if (complex.getImaginary() < 0 ) {
        out << "(" << complex.getReal() << " - " << -complex.getImaginary() << "i)";
        return out;
    } else {
    out << "(" << complex.getReal() << " + " << complex.getImaginary() << "i)";
    return out;
    }
}

File: WC3/Java_homework/Basic/Complex/Complex.java
public class Complex implements Cloneable {
    double real;
    double imaginary;
    
    public Complex(double real, double imaginary) {
        this.real = real;
        this.imaginary = imaginary;
    }
    public Complex() {
        this(0,0);
    }
    
    public double getReal() {
        return this.real;
    }
    
    public double getImaginary() {
        return this.imaginary;
    }
    
    public void setReal(double value) {
        this.real = value;
    }
    
    public void setImaginary(double value) {
        this.imaginary = value;
    }
    
    public boolean equals(Complex other) {
        return this.real == other.real && this.imaginary == other.imaginary;
    }
    
    public boolean notEqual(Complex other) {
        return !(this.equals(other));
    }
    
    public void increment(Complex other) {
        this.real += other.real;
        this.imaginary += other.imaginary;
    }
    
    public void decrement(Complex other) {
        this.real -= other.real;
        this.imaginary -= other.imaginary;
    }
    
    public void multiply(Complex other) {
        Complex multiple = new Complex();
        
        multiple.real = this.real * other.real - this.imaginary * other.imaginary;
        multiple.imaginary = this.imaginary * other.real + this.real * other.imaginary;
        this.real = multiple.real;
        this.imaginary = multiple.imaginary;
    }
    
    
    public Complex sum(Complex other) {
        Complex result = new Complex();
        
        result.real = this.real + other.real;
        result.imaginary = this.imaginary + other.imaginary;
        return result;
    }
    public Complex diff(Complex other) {
        Complex result = new Complex();
        
        result.real = this.real - other.real;
        result.imaginary = this.imaginary - other.imaginary;
        return result;
    }
    
    public String toString() {
        return "(" + this.real + " + " + this.imaginary + "i)";
    }
    
    public Complex clone() throws CloneNotSupportedException {
        return (Complex) super.clone();
    }
}
